File: neetcode/242. Valid Anagram/242.java
/*
Method 1.1:
Intuition: Compare their character frequency maps.
TC: O(N)
SC: O(1)
 */

class Solution {
    public boolean isAnagram(String s, String t) {
        int[] sMap = new int[26];

        for (char c : s.toCharArray()) {
            sMap[c - 'a']++;
        }

        for (char c : t.toCharArray()) {
            sMap[c - 'a']--;
        }

        for (int n : sMap) {
            if (n != 0) return false;
        }

        return true;
    }
}

File: neetcode/242. Valid Anagram/242.kt
/*
Method 1.0:
Intuition: Compare their character counts. Edge case: unequal lengths.
TC: O(N)
SC: O(1)
 */

class Solution {
    fun isAnagram(s: String, t: String): Boolean {
        val sMap = IntArray(26)

        if (s.length != t.length) return false

        for (c in s.toCharArray()) {
            sMap[c - 'a']++
        }

        for (c in t.toCharArray()) {
            if (--sMap[c - 'a'] < 0) return false
        }

        return true
    }
}
